src/ApparentHorizons/HorizonComponentActions.hpp

```diff
@@ -0,0 +1,388 @@
+    if (verbose > ::Verbosity::Quiet) {
+      Parallel::printf("FastFlowIteration: %s has converged, timestep %s\n",
+                       AhTag::label(), step);
+    }
+
+    // Tell all the DataInterpolators about the convergence.
+    // Eventually when there are Actions for observers, control
+    // systems, etc. that depend on the horizon, they should be called
+    // here.
+    auto& horizon_manager_proxy =
+        Parallel::get_parallel_component<ah::DataInterpolator<Metavariables>>(
+            cache);
+    Parallel::simple_action<
+        Actions::DataInterpolator::HorizonHasConverged<AhTag>>(
+        horizon_manager_proxy, step);
+
+    // Call the hook.
+    const auto& strahlkorper =
+        db::get<Tags::Strahlkorper<typename AhTag::frame>>(box);
+    AhTag::convergence_hook::apply(strahlkorper, step, cache);
+
+    // Prepare for finding horizon at a new time.
+    // For now, the initial guess for the new
+    // horizon is the result of the old one.
+    // Eventually ah::Finder will hold the previous horizons and
+    // will do time-extrapolation to set the next guess.
+    db::mutate<Tags::FastFlow, Tags::Timesteps>(
+        make_not_null(&box), [](const gsl::not_null<::FastFlow*> fast_flow,
+                                const gsl::not_null<
+                                    db::item_type<Tags::Timesteps>*>
+                                    steps) noexcept {
+          fast_flow->reset_for_next_find();
+          steps->pop_front();
+        });
+
+    const auto& timesteps = db::get<Tags::Timesteps>(box);
+    if (not timesteps.empty()) {
+      // There are still horizon searches to be done, so do the next one.
+      // Note that we haven't yet implemented time extrapolation for a
+      // new initial guess.  Currently the previous Strahlkorper
+      // (which is in the DataBox) is used for the initial guess of the
+      // next search.  To add the time extrapolation, we need to store
+      // one or more extra Strahlkorper(s) for the previous iteration(s),
+      // and do the extrapolation here.
+      send_points_to_horizon_manager<AhTag, DbTags, Metavariables>(
+          box, cache, timesteps.front());
+    }
+  } else if (status == FastFlow::Status::SuccessfulIteration) {
+    // Do another iteration of the same horizon search.
+    if (verbose == ::Verbosity::Debug) {
+      Parallel::printf(
+          "### Node:%d  Proc:%d ###\n"
+          "%s: FastFlowIteration: Horizon next iter %d\n\n",
+          Parallel::my_node(), Parallel::my_proc(), AhTag::label(),
+          info.iteration);
+    }
+    const auto& timesteps = db::get<Tags::Timesteps>(box);
+    send_points_to_horizon_manager<AhTag, DbTags, Metavariables>(
+        box, cache, timesteps.front());
+  } else {
+    Parallel::printf("Finder %s failed, reason = %s\n", AhTag::label(), status);
+    Parallel::abort("Finder failed");
+  }
+}
+}  // namespace Finder_detail
+
+namespace Actions {
```

Comment:
Please split the individual actions into separate files.  Test each action individually with the action testing framework.
Then you can add each action as an individual commit and the actions can be a separate PR from the components.